oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/osh/OSHEntityTimeUtils.java

```diff
@@ -0,0 +1,357 @@
+    Map<OSHDBTimestamp, Long> result = new TreeMap<>();
+
+    List<OSMWay> ways = org.heigit.ohsome.oshdb.osh.OSHEntities.toList(osh.getVersions());
+    ways.forEach(osm -> result.put(osm.getTimestamp(), osm.getChangesetId()));
+
+    // recurse way nodes
+    try {
+      for (OSHNode node : osh.getNodes()) {
+        if (node == null) {
+          continue;
+        }
+        for (OSMNode version : node.getVersions()) {
+          result.putIfAbsent(version.getTimestamp(), version.getChangesetId());
+        }
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    return result;
+  }
+
+  private static Map<OSHDBTimestamp, Long> getChangesetTimestamps(OSHRelation osh) {
+    Map<OSHDBTimestamp, Long> result = new TreeMap<>();
+
+    List<OSMRelation> rels = org.heigit.ohsome.oshdb.osh.OSHEntities.toList(osh.getVersions());
+    rels.forEach(osmRel -> result.put(osmRel.getTimestamp(), osmRel.getChangesetId()));
+
+    // recurse rel members
+    try {
+      Stream.concat(osh.getNodes().stream(), osh.getWays().stream())
+          .filter(Objects::nonNull)
+          .forEach(oshEntity -> getChangesetTimestamps(oshEntity).forEach(result::putIfAbsent));
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+
+    return result;
+  }
+
+  /**
+   * Returns all timestamps at which this entity (or one or more of its child entities) has been
+   * modified.
+   * 
+   * @param osh the osh entity to work on
+   * @return a list of timestamps where this entity has been modified
+   */
+  public static List<OSHDBTimestamp> getModificationTimestamps(OSHEntity osh) {
+    return getModificationTimestamps(osh, true);
+  }
+
+  /**
+   * Returns the list of timestamps at which this entity was modified.
+   *
+   * <p>
+   * If the parameter "recurse" is set to true, it will also include modifications of the object's
+   * child elements (useful to find out when the geometry of this object has been altered).
+   * </p>
+   *
+   * @param osh the osh entity to work on
+   * @param recurse specifies if times of modifications of child entities should also be returned or
+   *        not
+   * @return a list of timestamps where this entity has been modified
+   */
+  public static List<OSHDBTimestamp> getModificationTimestamps(OSHEntity osh, boolean recurse) {
+    switch (osh.getType()) {
+      case NODE:
+        return getModificationTimestamps((OSHNode) osh);
+      case WAY:
+      case RELATION:
+        return getModificationTimestamps(osh, recurse, null);
+      default:
+        throw new UnsupportedOperationException(format(UNSUPPORTED_OSMTYPE_MESSAGE, osh.getType()));
+    }
+  }
+
+  private static List<OSHDBTimestamp> getModificationTimestamps(OSHNode osh) {
+    List<OSHDBTimestamp> result = new ArrayList<>();
+    for (OSMEntity osm : osh.getVersions()) {
+      result.add(osm.getTimestamp());
+    }
```

Comment:
we should size the ArrayList to avoid growing. like it is done in `OSHEntities.toList`

```java
var itr = osh.getVersions().iterator();
if (!itr.hasNext()) {
  return Collections.emptyList();
}
var osm = itr.next();
var result = new ArrayList<OSHDBTimestamp>(osm.getVersion());
result.add(osm.getTimestamp());
while (itr.hasNext()) {
  osm = itr.next();
  result.add(osm.getTimestamp());
}
```
```

Reply:
good suggestion. I made `OSHEntities.toList` a bit more powerful to accommodate  for this use case as well.